avocado/utils/process.py

```diff
@@ -307,6 +307,9 @@ def __init__(self, command="", stdout=b"", stderr=b"",
         self.pid = pid
         if encoding is None:
             encoding = sys.getdefaultencoding()
+            if PY2 and encoding == "ascii":
+                # python2 fails to getdefaultencoding, let's assume utf-8
```

Comment:
You're missing the point about `getdefaultencoding`.  It has nothing to do with system environment variables, it is a constant set on Python source code.

We may add an **avocado** configuration for the encoding to be used, but this has nothing to with Python's own.

Reply:
Well the `getdefaultencoding()` was here before, I just add `utf-8` in case it reports `ascii`, because that is (known) py2 odd behavior. Any suggestions how to do this better? I don't think we should make this configurable at this point, we can always add it if users ask for it.

Reply:
The fact that you're switching to another encoding, one that the user did not ask for, feels wrong.  I mean, as the user of an API documented as:

```
    :param encoding: the encoding to use for the text version
                     of stdout and stderr, with the default being
                     Python's own (:func:`sys.getdefaultencoding`).
```

Having anything different than that would raise serious "WTHs" from any caring users.  Additionally, the following statement is incorrect, and makes me thing that you're missing the point about `getdefaultencoding()`:

"# python2 fails to getdefaultencoding, let's assume utf-8"

I've never seen Python 2 failing to get the default encoding, because the default encoding is a static string compiled into the Python interpreter, and not something that it gets from the system or user settings.

The real, and harder solution here, seems to be the one yourself coined: having a central configuration/function to encode/decode strings.  I'd be fine with hard coded utf-8 encoding in Python 2.  Then we need to make sure all input taken (including this case, commands to be executed, are treated as such).  If API users have data in different encodings, it'd up to then to turn it into utf-8 before feeding the data into Avocado.

Comments?

Reply:
Yes, I'd prefer https://trello.com/c/wsd2z7WS/1265-unify-py3-py2-default-encoding-handling and when that is available I'd happily use it here as well. But so far we don't have anything like that and many tests in my CI are failing because of the encoding handling added to support python3.

So this is a bugfix, it hardcodes what is suggested out there (I did not made this up, I saw it on multiple places where they mentioned the differences between py2 and py3 and that this function just never works on py2). So how about merging this to avoid the breakage and adding task to https://trello.com/c/wsd2z7WS/1265-unify-py3-py2-default-encoding-handling to use it from here?

Reply:
I'm not against defining a new default encoding type for Python 2 here, but there a few broken issues with the code AS IS:

1) It makes no sense to do `sys.getdefaultencoding()`, and check for `ascii` and then set to `utf-8` because on Python 2, `sys.getdefaultencoding()` will **always** return `ascii`.  It's hardcoded into the Python source code:  https://github.com/python/cpython/blob/2.7/Objects/unicodeobject.c#L126

2) The documentation of the `CmdResult` class would need an update

The ideal usage would be something like:

```python
# this is Python 2.7
result = process.run('cmd_that_generates_bytes_larger_than_127')
result.encoding = 'utf-8'
process_text(result.stdout)
```

But I agree that sometimes (when running a simple test for instance), we can't do `result.encoding = 'utf-8'`.  That's why I still think we should fix this issue for good, setting a comprehensive default encoding (such as utf-8).

Does it make sense?